File: data_structure/FenwickTree/cpp/FenwickTree.cpp
#include "FenwickTree.hpp"

#include <limits>

using namespace std;

namespace {
bool CheckedAdd(long long a, long long b, long long& result) {
    if ((b > 0 && a > numeric_limits<long long>::max() - b) ||
        (b < 0 && a < numeric_limits<long long>::min() - b)) {
        return false;
    }
    result = a + b;
    return true;
}

bool CheckedSub(long long a, long long b, long long& result) {
    if (b == numeric_limits<long long>::min()) {
        return false;
    }
    return CheckedAdd(a, -b, result);
}

class FenwickTree {
public:
    explicit FenwickTree(int n) : tree_(n + 1, 0) {}

    bool Add(int index, long long delta) {
        for (int i = index + 1; i < static_cast<int>(tree_.size()); i += i & -i) {
            long long updated = 0;
            if (!CheckedAdd(tree_[i], delta, updated)) {
                return false;
            }
            tree_[i] = updated;
        }
        return true;
    }

    bool PrefixSum(int index, long long& result) const {
        result = 0;
        for (int i = index + 1; i > 0; i -= i & -i) {
            long long updated = 0;
            if (!CheckedAdd(result, tree_[i], updated)) {
                return false;
            }
            result = updated;
        }
        return true;
    }

    bool RangeSum(int left, int right, long long& result) const {
        long long rightPrefix = 0;
        if (!PrefixSum(right, rightPrefix)) {
            return false;
        }

        if (left == 0) {
            result = rightPrefix;
            return true;
        }

        long long leftPrefix = 0;
        if (!PrefixSum(left - 1, leftPrefix)) {
            return false;
        }

        return CheckedSub(rightPrefix, leftPrefix, result);
    }

private:
    vector<long long> tree_;
};
}  // namespace

vector<long long> ProcessFenwickQueries(
    const vector<long long>& initialValues,
    const vector<Query>& queries,
    string& errorMessage
) {
    FenwickTree tree(static_cast<int>(initialValues.size()));
    vector<long long> results;

    for (int index = 0; index < static_cast<int>(initialValues.size()); ++index) {
        if (!tree.Add(index, initialValues[index])) {
            errorMessage = "Fenwick tree overflowed while building the initial state.";
            return {};
        }
    }

    results.reserve(queries.size());
    for (const Query& query : queries) {
        if (query.type == QueryType::Add) {
            if (!tree.Add(query.left, query.delta)) {
                errorMessage = "Fenwick tree overflowed while applying an update.";
                return {};
            }
        } else {
            long long sum = 0;
            if (!tree.RangeSum(query.left, query.right, sum)) {
                errorMessage = "Fenwick tree overflowed while evaluating a sum query.";
                return {};
            }
            results.push_back(sum);
        }
    }

    return results;
}

File: data_structure/FenwickTree/rust/src/fenwick_tree.rs
#[derive(Clone, Copy)]
pub enum Query {
    Add { index: usize, delta: i64 },
    Sum { left: usize, right: usize },
}

pub struct FenwickTree {
    tree: Vec<i64>,
}

impl FenwickTree {
    pub fn new(size: usize) -> Self {
        Self { tree: vec![0; size + 1] }
    }

    pub fn add(&mut self, index: usize, delta: i64) -> Result<(), String> {
        let mut i = index + 1;
        while i < self.tree.len() {
            self.tree[i] = self.tree[i]
                .checked_add(delta)
                .ok_or_else(|| String::from("Fenwick tree overflowed while applying an update."))?;
            i += i & (!i + 1);
        }
        Ok(())
    }

    pub fn prefix_sum(&self, index: usize) -> Result<i64, String> {
        let mut result = 0_i64;
        let mut i = index + 1;
        while i > 0 {
            result = result
                .checked_add(self.tree[i])
                .ok_or_else(|| String::from("Fenwick tree overflowed while evaluating a sum query."))?;
            i -= i & (!i + 1);
        }
        Ok(result)
    }

    pub fn range_sum(&self, left: usize, right: usize) -> Result<i64, String> {
        let right_prefix = self.prefix_sum(right)?;
        if left == 0 {
            return Ok(right_prefix);
        }
        let left_prefix = self.prefix_sum(left - 1)?;
        right_prefix
            .checked_sub(left_prefix)
            .ok_or_else(|| String::from("Fenwick tree overflowed while evaluating a sum query."))
    }
}

pub fn process_fenwick_queries(initial_values: &[i64], queries: &[Query]) -> Result<Vec<i64>, String> {
    let mut tree = FenwickTree::new(initial_values.len());
    for (index, value) in initial_values.iter().enumerate() {
        tree.add(index, *value)
            .map_err(|_| String::from("Fenwick tree overflowed while building the initial state."))?;
    }

    let mut results = Vec::new();
    for query in queries {
        match *query {
            Query::Add { index, delta } => tree.add(index, delta)?,
            Query::Sum { left, right } => results.push(tree.range_sum(left, right)?),
        }
    }

    Ok(results)
}
